extensions/consul-config/runtime/src/main/java/io/quarkus/consul/config/runtime/ConsulConfig.java

```diff
@@ -0,0 +1,117 @@
+package io.quarkus.consul.config.runtime;
+
+import java.net.InetSocketAddress;
+import java.time.Duration;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+import io.quarkus.runtime.annotations.ConfigGroup;
+import io.quarkus.runtime.annotations.ConfigItem;
+import io.quarkus.runtime.annotations.ConfigPhase;
+import io.quarkus.runtime.annotations.ConfigRoot;
+
+@ConfigRoot(name = "consul-config", phase = ConfigPhase.BOOTSTRAP)
+public class ConsulConfig {
+
+    /**
+     * If set to true, the application will attempt to look up the configuration from Consul
+     */
+    @ConfigItem(defaultValue = "false")
+    boolean enabled;
+
+    /**
+     * Consul agent related configuration
+     */
+    @ConfigItem
+    AgentConfig agent;
+
+    /**
+     * Common prefix that all keys share when looking up the keys from Consul.
+     * The prefix is <b>not</b> included in the keys of the user configuration
+     */
+    @ConfigItem
+    Optional<String> prefix;
+
+    /**
+     * Keys whose value is a raw string.
+     * When this is used, the keys that end up in the user configuration are the keys specified her with '/' replaced by '.'
+     */
+    @ConfigItem
+    Optional<List<String>> rawValueKeys;
+
+    /**
+     * Keys whose value represents a properties file.
+     * When this is used, the keys that end up in the user configuration are the keys of the of the properties file,
+     * not these keys
+     */
+    @ConfigItem
+    Optional<List<String>> propertiesValueKeys;
+
+    /**
+     * If set to true, the application will not start if any of the configured config sources cannot be located
+     */
+    @ConfigItem(defaultValue = "true")
+    boolean failOnMissingKey;
+
+    Map<String, ValueType> keysAsMap() {
+        Map<String, ValueType> result = new LinkedHashMap<>();
+        if (rawValueKeys.isPresent()) {
+            for (String key : rawValueKeys.get()) {
+                result.put(key, ValueType.RAW);
+            }
+        }
+        if (propertiesValueKeys.isPresent()) {
+            for (String key : propertiesValueKeys.get()) {
+                result.put(key, ValueType.PROPERTIES);
+            }
+        }
+        return result;
+    }
+
+    @ConfigGroup
+    public static class AgentConfig {
+
+        /**
+         * Consul agent host
+         */
+        @ConfigItem(defaultValue = "localhost:8500")
+        InetSocketAddress hostPort;
+
+        /**
+         * Whether or not to use HTTPS when communicating with the agent
+         */
+        @ConfigItem(defaultValue = "false")
+        boolean useHttps;
+
+        /**
+         * Consul token to be provided when authentication is enabled
+         */
+        @ConfigItem
+        Optional<String> token;
+
+        /**
+         * When using HTTPS, whether or not to trust self-signed certificatd
+         */
+        @ConfigItem(defaultValue = "false")
+        boolean trustCerts;
```

Comment:
Don't we need a trustStore configuration? Also might be a good idea to check other extensions for consistency on this.

Reply:
Yeah, I was planning on looking into it

Reply:
I wonder if we should have some sort of shared config group for that. That would enforce consistency.

Reply:
That's what Max was saying IIRC in a mailing thread.

I am personally sceptical, but not completely opposed to it.

Reply:
I added support for a KeyStore file